File: src/main/java/org/simbrain/workspace/serialization/WorkspaceSerializer.java
/*
 * Part of Simbrain--a java-based neural network kit
 * Copyright (C) 2005,2007 The Authors.  See http://www.simbrain.net/credits
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package org.simbrain.workspace.serialization;

import org.simbrain.util.SFileChooser;
import org.simbrain.util.SimbrainPreferences;
import org.simbrain.util.XStreamUtils;
import org.simbrain.workspace.*;
import org.simbrain.workspace.couplings.Coupling;
import org.simbrain.workspace.gui.DesktopComponent;
import org.simbrain.workspace.gui.SimbrainDesktop;
import org.simbrain.workspace.updater.UpdateAction;
import org.simbrain.workspace.updater.UpdateActionManager;

import javax.swing.*;
import java.awt.*;
import java.io.*;
import java.lang.reflect.Method;
import java.util.List;
import java.util.*;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;
import java.util.zip.ZipOutputStream;

/**
 * Serializes and deserializes workspaces. Custom serialization (beyond what XStream can do) is required, in order to
 * recreate workspace components and couplings from a legible xml form in a zipped directory structure. Some additional
 * information is in {@link Workspace}.
 * <br>
 * To make a workspace component serializable, here are some things you should do.
 * <ol>
 * <li>In your subclass of {@link WorkspaceComponent}, override {@link WorkspaceComponent#getXML()} and
 * {@link WorkspaceComponent#save(OutputStream, String)}}, which convert your main model objects to and from xml.</li>
 * <li>In your subclass of {@link WorkspaceComponent}, override {@link WorkspaceComponent#getAttributeContainers()}.
 * It should return a list of objects that contain consumers or producers, i.e. that contain methods
 * annotated as {@link Consumable} or {@link Producible}.</li>
 * <li>Mark any fields that should not be serialized as transient.</li>
 * <li>Special initialization when de-serializing can be handled with ReadResolve methods, or by using custom XStream
 * converters. For details on de-serialization see {@link WorkspaceComponentDeserializer}</li>
 * </ol>
 * <br>
 * Some good examples are in {@link org.simbrain.network.NetworkComponent} and {@link
 * org.simbrain.world.odorworld.OdorWorldComponent}.
 * <br>
 * @author Matt Watson
 * @author Zoë Tosi
 * @author Jeff Yoshimi
 */
public class WorkspaceSerializer {

    /**
     * The number of bytes to attempt to read at a time from an InputStream.
     */
    private static final int BUFFER_SIZE = 1024;

    /**
     * The current workspace.
     */
    private Workspace workspace;

    /**
     * The desktop component for the workspace.
     */
    private SimbrainDesktop desktop;

    /**
     * Creates a new serializer.
     *
     * @param workspace The workspace to serialize to or from.
     */
    public WorkspaceSerializer(Workspace workspace) {
        this.workspace = workspace;
        this.desktop = SimbrainDesktop.INSTANCE;
    }

    /**
     * Serializes the workspace to a zip compressed stream.
     *
     * @param output The output stream to write to.
     * @throws IOException If there is an IO error.
     */
    public void serialize(OutputStream output) throws IOException {

        // Create the zip output stream. ZipStream is a sequence of
        // ZipEntries, with extra utilities for iterating over them.
        // Each zipentry corresponds to a single file in the zip archive, a
        // String with the relative path in the archive to the entry (e.g.
        // "gui/network.xml"), and a bytearray for the file itself.
        ZipOutputStream zipStream = new ZipOutputStream(output);
        WorkspaceComponentSerializer serializer = new WorkspaceComponentSerializer();

        // This archive object saves all the information about the workspace. It
        // will be saved as a zipentry "contents.xml"
        ArchivedWorkspace archive = new ArchivedWorkspace(workspace, serializer);

        // Currently sorts components by a serialization priority
        workspace.preSerializationInit();

        serializeComponents(serializer, archive, zipStream);
        serializeCouplings(archive);
        // serializeUpdateActions(archive);

        ZipEntry entry = new ZipEntry("contents.xml");
        zipStream.putNextEntry(entry);
        archive.toXml(zipStream);
        zipStream.finish();
    }

    /**
     * Serializes all the components to the given archive and zipstream.
     *
     * @param serializer The serializer for the components.
     * @param archive    The archive contents to update.
     * @param zipStream  The zipstream to write to.
     * @throws IOException If there is an IO error.
     */
    private void serializeComponents(WorkspaceComponentSerializer serializer, ArchivedWorkspace archive, ZipOutputStream zipStream) throws IOException {
        List<WorkspaceComponent> components = sortComponentsByPriority();
        for (WorkspaceComponent component : workspace.getComponentList()) {
            serializeComponent(serializer, archive, component, zipStream);
        }
    }

    private List<WorkspaceComponent> sortComponentsByPriority() {
        List<WorkspaceComponent> components = new ArrayList<WorkspaceComponent>();
        components.addAll(workspace.getComponentList());
        Collections.sort(components, new Comparator<WorkspaceComponent>() {
            public int compare(WorkspaceComponent c1, WorkspaceComponent c2) {
                return Integer.compare(c1.getSerializePriority(), c2.getSerializePriority());
            }
        });
        return components;
    }

    /**
     * Serialize one component to the zip stream
     *
     * @param serializer The serializer for the components.
     * @param archive    The archive contents to update.
     * @param component  the component to serialize
     * @param zipStream  The zipstream to write to.
     */
    private void serializeComponent(WorkspaceComponentSerializer serializer, ArchivedWorkspace archive, WorkspaceComponent component, ZipOutputStream zipStream) {
        ArchivedWorkspaceComponent archiveComp = archive.addComponent(component);
        ZipEntry entry = new ZipEntry(archiveComp.getUri());
        try {
            zipStream.putNextEntry(entry);
            serializer.serializeComponent(component, zipStream);
            if (SimbrainDesktop.INSTANCE != null) {
                DesktopComponent<?> desktopComponent = SimbrainDesktop.INSTANCE.getDesktopComponent(component);
                // Makes it possible to save a non-GUI simulation
                if (desktopComponent != null) {
                    ArchivedWorkspaceComponent.ArchivedDesktopComponent dc = archiveComp.addDesktopComponent(desktopComponent);
                    entry = new ZipEntry(dc.getUri());
                    zipStream.putNextEntry(entry);
                    desktopComponent.save(zipStream);
                }
            }
        } catch (IOException ex) {
            ex.printStackTrace();
        }
    }

    /**
     * Serialize couplings.
     *
     * @param archive the archive objet to serialize to
     */
    private void serializeCouplings(ArchivedWorkspace archive) {
        HashMap<Object, WorkspaceComponent> couplingComponents = mapCouplingComponents();
        for (Coupling coupling : workspace.getCouplings()) {
            serializeCoupling(couplingComponents, coupling, archive);
        }
    }

    /**
     * Returns a map from coupling base objects ("models") to their parent components.
     */
    private HashMap<Object, WorkspaceComponent> mapCouplingComponents() {
        HashMap<Object, WorkspaceComponent> couplingComponents = new HashMap<>();
        for (WorkspaceComponent component : workspace.getComponentList()) {
            for (Object object : component.getAttributeContainers()) {
                couplingComponents.put(object, component);
            }
        }
        return couplingComponents;
    }

    /**
     * Add a serialized coupling to the archive.
     *
     * @param couplingComponents a map from couplings to components
     * @param coupling           the coupling to save
     * @param archive            the archive object to save to
     */
    private void serializeCoupling(HashMap<Object, WorkspaceComponent> couplingComponents, Coupling coupling, ArchivedWorkspace archive) {
        ArchivedAttribute producer = new ArchivedAttribute(couplingComponents.get(coupling.getProducer().getBaseObject()), coupling.getProducer());
        ArchivedAttribute consumer = new ArchivedAttribute(couplingComponents.get(coupling.getConsumer().getBaseObject()), coupling.getConsumer());
        archive.addCoupling(new ArchivedCoupling(producer, consumer));
    }

    /**
     * Serialize all update actions  in the current workspace
     *
     * @param archive the archive object to serialize to
     */
    private void serializeUpdateActions(ArchivedWorkspace archive) {
        for (UpdateAction action : workspace.getUpdater().getUpdateManager().getActionList()) {
            archive.addUpdateAction(action);
        }
    }

    /**
     * Creates a workspace from a zip compressed input stream.
     *
     * @param stream The stream to read from. This is expected to be zip compressed.
     * @throws IOException if an IO error occurs.
     */
    @SuppressWarnings("unchecked")
    public void deserialize(InputStream stream) throws IOException {
        Map<String, byte[]> byteArrays = processInputStream(stream);
        ArchivedWorkspace archive = (ArchivedWorkspace) ArchivedWorkspace.xstream().fromXML(new ByteArrayInputStream(byteArrays.get("contents.xml")));

        WorkspaceComponentDeserializer deserializer = new WorkspaceComponentDeserializer();
        deserializeComponents(archive, deserializer, byteArrays);

        deserializeCouplings(archive);
        // deserializeUpdateActions(archive, deserializer);
        deserializeWorkspaceParameters(archive);

    }

    private Map<String, byte[]> processInputStream(InputStream stream) throws IOException {
        // Populate the byte stream BUFFER_SIZE at a time and create a zip input
        // stream (currently 1 kb at a time).
        ByteArrayOutputStream bytes = new ByteArrayOutputStream();
        byte[] buffer = new byte[BUFFER_SIZE];
        for (int read; (read = stream.read(buffer)) >= 0; ) {
            bytes.write(buffer, 0, read);
        }
        ZipInputStream zip = new ZipInputStream(new ByteArrayInputStream(bytes.toByteArray()));

        // Populate a map from zip entries (strings containing path+file info in
        // zip archive) to the associated data
        Map<String, byte[]> byteArrays = new HashMap<String, byte[]>();
        ZipEntry entry = zip.getNextEntry();
        // Initialize byte arrays
        for (ZipEntry next; entry != null; entry = next) {
            next = zip.getNextEntry();
            byteArrays.put(entry.getName(), new byte[(int) entry.getSize()]);
        }
        zip = new ZipInputStream(new ByteArrayInputStream(bytes.toByteArray()));
        // Populate byte arrays
        while ((entry = zip.getNextEntry()) != null) {
            byte[] data = byteArrays.get(entry.getName());
            read(zip, data);
        }

        // Find the contents.xml file and set the zip entries relative to that
        Set<String> zipEntries = new HashSet<String>(byteArrays.keySet());
        String contentsFile = "contents.xml";
        String contentsPath = "";
        for (String entryName : zipEntries) {
            if (entryName.endsWith(contentsFile)) {
                contentsPath = entryName.substring(0, entryName.length() - contentsFile.length());
            }
        }

        // Remove the contents path from all entries that have it
        if (!contentsPath.isEmpty()) {
            for (String entryName : zipEntries) {
                if (entryName.startsWith(contentsPath)) {
                    byteArrays.put(entryName.replace(contentsPath, ""), byteArrays.get(entryName));
                    byteArrays.remove(entryName);
                }
            }
        }
        return byteArrays;
    }

    private void deserializeComponents(ArchivedWorkspace archive, WorkspaceComponentDeserializer deserializer, Map<String, byte[]> byteArrays) {
        if (archive.getArchivedComponents() != null) {
            for (ArchivedWorkspaceComponent archivedComponent : archive.getArchivedComponents()) {
                try {
                    WorkspaceComponent wc = deserializer.deserializeWorkspaceComponent(archivedComponent, new ByteArrayInputStream(byteArrays.get(archivedComponent.getUri())));
                    workspace.addWorkspaceComponent(wc);
                    if (archivedComponent.getDesktopComponent() != null) {
                        Rectangle bounds =
                                (Rectangle) XStreamUtils.getSimbrainXStream().fromXML(new ByteArrayInputStream(byteArrays.get(archivedComponent.getDesktopComponent().getUri())));
                        DesktopComponent<?> desktopComponent = desktop.getDesktopComponent(wc);
                        desktopComponent.parentFrame.setBounds(bounds);
                    }
                } catch (Exception ex) {
                    ex.printStackTrace();
                    String message = String.format("Failed to deserialize component %s.", archivedComponent.getName());
                    JOptionPane.showMessageDialog(null, message);
                }
            }
        }
    }

    private void deserializeCouplings(ArchivedWorkspace archive) {
        if (archive.getArchivedCouplings() != null) {
            for (ArchivedCoupling archivedCoupling : archive.getArchivedCouplings()) {
                Producer producer = archivedCoupling.createProducer(workspace);
                Consumer consumer = archivedCoupling.createConsumer(workspace);
                workspace.getCouplingManager().createCoupling(producer, consumer);
            }
        }
    }

    private void deserializeUpdateActions(ArchivedWorkspace archive, WorkspaceComponentDeserializer deserializer) {
        UpdateActionManager manager = workspace.getUpdater().getUpdateManager();
        manager.clear();
        if (archive.getArchivedActions() != null) {
            for (ArchivedUpdateAction archivedAction : archive.getArchivedActions()) {
                manager.addAction(archive.createUpdateAction(workspace, deserializer, archivedAction));
            }
        }
    }

    private void deserializeWorkspaceParameters(ArchivedWorkspace archive) {
        if (archive.getWorkspaceParameters() != null) {
            workspace.setUpdateDelay(archive.getWorkspaceParameters().getUpdateDelay());
            workspace.getUpdater().setTime(archive.getWorkspaceParameters().getSavedTime());
            workspace.initIdManager();
        }
    }

    /**
     * Helper method that will read the InputStream repeatedly until the given array is filled.
     *
     * @param istream the InputStream to read from.
     * @param bytes   the array to write to
     * @throws IOException if there is an IO error
     */
    private static void read(InputStream istream, byte[] bytes) throws IOException {
        int pos = 0;
        while (pos < bytes.length) {
            int read = istream.read(bytes, pos, bytes.length - pos);
            if (read < 0) {
                throw new RuntimeException("premature EOF");
            }
            pos += read;
        }
    }

    public static <T> WorkspaceComponent showOpenComponentDialog(Class<T> type) {
        String defaultDirectory = SimbrainPreferences.getString("workspace" + type.getSimpleName() + "Directory");
        SFileChooser chooser = new SFileChooser(defaultDirectory, "XML File", "xml");
        File file = chooser.showOpenDialog();
        if (file != null) {
            return WorkspaceSerializer.open(type, file);
        } else {
            return null;
        }
    }

    /**
     * Helper method for openings workspace components from a file.
     * <p>
     * A call might look like this <code>NetworkComponent networkComponent = (NetworkComponent)
     * WorkspaceFileOpener(NetworkComponent.class, new File("Net.xml"));</code>
     *
     * @param fileClass the type of Workpsace component to open; a subclass of WorkspaceComponent.
     * @param file      the File to open
     * @return the workspace component
     */
    public static WorkspaceComponent open(Class<?> fileClass, File file) {
        String extension = file.getName().substring(file.getName().indexOf("."));
        try {
            Method method = fileClass.getMethod("open", InputStream.class, String.class, String.class);
            WorkspaceComponent wc = (WorkspaceComponent) method.invoke(null, new FileInputStream(file),
                    file.getName(), extension);
            wc.setCurrentFile(file);
            wc.setChangedSinceLastSave(false);
            return wc;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }



}

File: src/main/kotlin/org/simbrain/workspace/serialization/WorkspaceSerializer.kt
package org.simbrain.workspace.serialization

import org.simbrain.util.SFileChooser
import org.simbrain.util.getSimbrainXStream
import org.simbrain.workspace.Workspace
import org.simbrain.workspace.WorkspaceComponent
import org.simbrain.workspace.WorkspacePreferences.baseDirectory
import org.simbrain.workspace.couplings.Coupling
import org.simbrain.workspace.gui.SimbrainDesktop.getDesktopComponent
import java.awt.Rectangle
import java.io.*
import java.util.*
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream
import java.util.zip.ZipOutputStream
import javax.swing.JOptionPane

class WorkspaceSerializer(val workspace: Workspace) {

    /**
     * Serializes the workspace to a zip compressed stream.
     *
     * @param output The output stream to write to.
     * @throws IOException If there is an IO error.
     */
    @JvmOverloads
    fun serialize(output: OutputStream, headless: Boolean = false) {
        // Create the zip output stream. ZipStream is a sequence of
        // ZipEntries, with extra utilities for iterating over them.
        // Each zipentry corresponds to a single file in the zip archive, a
        // String with the relative path in the archive to the entry (e.g.
        // "gui/network.xml"), and a bytearray for the file itself.

        val zipStream = ZipOutputStream(output)
        val serializer = WorkspaceComponentSerializer()

        // This archive object saves all the information about the workspace. It
        // will be saved as a zipentry "contents.xml"
        val archive = ArchivedWorkspace(workspace, serializer)

        // Currently sorts components by a serialization priority
        workspace.preSerializationInit()

        serializeComponents(serializer, archive, zipStream, headless)
        serializeCouplings(archive)

        // serializeUpdateActions(archive);
        val entry = ZipEntry("contents.xml")
        zipStream.putNextEntry(entry)
        archive.toXml(zipStream)
        zipStream.finish()
    }

    /**
     * Serializes all the components to the given archive and zipstream.
     *
     * @param serializer The serializer for the components.
     * @param archive    The archive contents to update.
     * @param zipStream  The zipstream to write to.
     * @throws IOException If there is an IO error.
     */
    @Throws(IOException::class)
    private fun serializeComponents(
        serializer: WorkspaceComponentSerializer,
        archive: ArchivedWorkspace,
        zipStream: ZipOutputStream,
        headless: Boolean
    ) {
        val components = sortComponentsByPriority()
        for (component in workspace.componentList) {
            serializeComponent(serializer, archive, component, zipStream, headless)
        }
    }

    private fun sortComponentsByPriority(): List<WorkspaceComponent> {
        val components: MutableList<WorkspaceComponent> = ArrayList()
        components.addAll(workspace.componentList)
        components.sortWith { c1, c2 -> c1.serializePriority.compareTo(c2.serializePriority) }
        return components
    }

    /**
     * Serialize one component to the zip stream
     *
     * @param serializer The serializer for the components.
     * @param archive    The archive contents to update.
     * @param component  the component to serialize
     * @param zipStream  The zipstream to write to.
     */
    private fun serializeComponent(
        serializer: WorkspaceComponentSerializer,
        archive: ArchivedWorkspace,
        component: WorkspaceComponent,
        zipStream: ZipOutputStream,
        headless: Boolean
    ) {
        val archiveComp = archive.addComponent(component)
        var entry: ZipEntry? = ZipEntry(archiveComp.getUri())
        try {
            zipStream.putNextEntry(entry)
            serializer.serializeComponent(component, zipStream)
            if (!headless) {
                val desktopComponent = getDesktopComponent(component)
                // Makes it possible to save a non-GUI simulation
                if (desktopComponent != null) {
                    val dc = archiveComp.addDesktopComponent(desktopComponent)
                    entry = ZipEntry(dc.uri)
                    zipStream.putNextEntry(entry)
                    desktopComponent.save(zipStream)
                }
            }
        } catch (ex: IOException) {
            ex.printStackTrace()
        }
    }

    /**
     * Serialize couplings.
     *
     * @param archive the archive objet to serialize to
     */
    private fun serializeCouplings(archive: ArchivedWorkspace) {
        val couplingComponents = mapCouplingComponents()
        for (coupling in workspace.couplings) {
            serializeCoupling(couplingComponents, coupling, archive)
        }
    }

    /**
     * Returns a map from coupling base objects ("models") to their parent components.
     */
    private fun mapCouplingComponents(): HashMap<Any, WorkspaceComponent> {
        val couplingComponents = HashMap<Any, WorkspaceComponent>()
        for (component in workspace.componentList) {
            for (`object` in component.attributeContainers) {
                couplingComponents[`object`] = component
            }
        }
        return couplingComponents
    }

    /**
     * Add a serialized coupling to the archive.
     *
     * @param couplingComponents a map from couplings to components
     * @param coupling           the coupling to save
     * @param archive            the archive object to save to
     */
    private fun serializeCoupling(
        couplingComponents: HashMap<Any, WorkspaceComponent>,
        coupling: Coupling,
        archive: ArchivedWorkspace
    ) {
        val producer = ArchivedAttribute(couplingComponents[coupling.producer.baseObject], coupling.producer)
        val consumer = ArchivedAttribute(couplingComponents[coupling.consumer.baseObject], coupling.consumer)
        archive.addCoupling(ArchivedCoupling(producer, consumer))
    }

    /**
     * Serialize all update actions  in the current workspace
     *
     * @param archive the archive object to serialize to
     */
    private fun serializeUpdateActions(archive: ArchivedWorkspace) {
        for (action in workspace.updater.updateManager.actionList) {
            archive.addUpdateAction(action)
        }
    }

    /**
     * Creates a workspace from a zip compressed input stream.
     *
     * @param stream The stream to read from. This is expected to be zip compressed.
     * @throws IOException if an IO error occurs.
     */
    @Throws(IOException::class)
    fun deserialize(stream: InputStream) {
        val byteArrays = processInputStream(stream)
        val archive =
            ArchivedWorkspace.xstream().fromXML(ByteArrayInputStream(byteArrays["contents.xml"])) as ArchivedWorkspace

        val deserializer = WorkspaceComponentDeserializer()
        deserializeComponents(archive, deserializer, byteArrays)

        deserializeCouplings(archive)
        // deserializeUpdateActions(archive, deserializer);
        deserializeWorkspaceParameters(archive)
    }

    @Throws(IOException::class)
    private fun processInputStream(stream: InputStream): Map<String, ByteArray?> {
        // Populate the byte stream BUFFER_SIZE at a time and create a zip input
        // stream (currently 1 kb at a time).
        val bytes = ByteArrayOutputStream()
        val buffer = ByteArray(BUFFER_SIZE)
        var read: Int
        while ((stream.read(buffer).also { read = it }) >= 0) {
            bytes.write(buffer, 0, read)
        }
        var zip = ZipInputStream(ByteArrayInputStream(bytes.toByteArray()))

        // Populate a map from zip entries (strings containing path+file info in
        // zip archive) to the associated data
        val byteArrays: MutableMap<String, ByteArray> = HashMap()
        var entry = zip.nextEntry
        // Initialize byte arrays
        var next: ZipEntry?
        while (entry != null) {
            next = zip.nextEntry
            byteArrays[entry.name] = ByteArray(entry.size.toInt())
            entry = next
        }
        zip = ZipInputStream(ByteArrayInputStream(bytes.toByteArray()))
        // Populate byte arrays
        while ((zip.nextEntry.also { entry = it }) != null) {
            val data = byteArrays[entry!!.name]!!
            read(zip, data)
        }

        // Find the contents.xml file and set the zip entries relative to that
        val zipEntries: Set<String> = HashSet(byteArrays.keys)
        val contentsFile = "contents.xml"
        var contentsPath = ""
        for (entryName in zipEntries) {
            if (entryName.endsWith(contentsFile)) {
                contentsPath = entryName.substring(0, entryName.length - contentsFile.length)
            }
        }

        // Remove the contents path from all entries that have it
        if (!contentsPath.isEmpty()) {
            for (entryName in zipEntries) {
                if (entryName.startsWith(contentsPath)) {
                    byteArrays[entryName.replace(contentsPath, "")] = byteArrays[entryName]!!
                    byteArrays.remove(entryName)
                }
            }
        }
        return byteArrays
    }

    private fun deserializeComponents(
        archive: ArchivedWorkspace,
        deserializer: WorkspaceComponentDeserializer,
        byteArrays: Map<String, ByteArray?>
    ) {
        if (archive.archivedComponents != null) {
            for (archivedComponent in archive.archivedComponents) {
                try {
                    val wc = deserializer.deserializeWorkspaceComponent(
                        archivedComponent,
                        ByteArrayInputStream(byteArrays[archivedComponent.getUri()])
                    )
                    wc.postOpenInit(workspace)
                    workspace.addWorkspaceComponent(wc)
                    if (archivedComponent.desktopComponent != null) {
                        val bounds =
                            getSimbrainXStream().fromXML(ByteArrayInputStream(byteArrays[archivedComponent.desktopComponent.uri])) as Rectangle
                        val desktopComponent = getDesktopComponent(wc)
                        desktopComponent.parentFrame.bounds = bounds
                    }
                } catch (ex: Exception) {
                    ex.printStackTrace()
                    val message = String.format("Failed to deserialize component %s.", archivedComponent.name)
                    JOptionPane.showMessageDialog(null, message)
                }
            }
        }
    }

    private fun deserializeCouplings(archive: ArchivedWorkspace) {
        if (archive.archivedCouplings != null) {
            for (archivedCoupling in archive.archivedCouplings) {
                val producer = archivedCoupling.createProducer(workspace)
                val consumer = archivedCoupling.createConsumer(workspace)
                workspace.couplingManager.createCoupling(producer, consumer, false)
            }
        }
    }

    private fun deserializeUpdateActions(archive: ArchivedWorkspace, deserializer: WorkspaceComponentDeserializer) {
        val manager = workspace.updater.updateManager
        manager.clear()
        if (archive.archivedActions != null) {
            for (archivedAction in archive.archivedActions) {
                manager.addAction(archive.createUpdateAction(workspace, deserializer, archivedAction))
            }
        }
    }

    private fun deserializeWorkspaceParameters(archive: ArchivedWorkspace) {
        if (archive.workspaceParameters != null) {
            workspace.updateDelay = archive.workspaceParameters.updateDelay
            workspace.updater.time = archive.workspaceParameters.savedTime
            workspace.initIdManager()
        }
    }

    companion object {
        /**
         * Helper method that will read the InputStream repeatedly until the given array is filled.
         *
         * @param istream the InputStream to read from.
         * @param bytes   the array to write to
         * @throws IOException if there is an IO error
         */
        @Throws(IOException::class)
        fun read(istream: InputStream, bytes: ByteArray) {
            var pos = 0
            while (pos < bytes.size) {
                val read = istream.read(bytes, pos, bytes.size - pos)
                if (read < 0) {
                    throw RuntimeException("premature EOF")
                }
                pos += read
            }
        }

        fun <T> showOpenComponentDialog(type: Class<T>): WorkspaceComponent? {
            val defaultDirectory = baseDirectory
            val chooser = SFileChooser(defaultDirectory, "XML File", "xml")
            val file = chooser.showOpenDialog()
            return if (file != null) {
                open(type, file)
            } else {
                null
            }
        }

        /**
         * Helper method for openings workspace components from a file.
         *
         *
         * A call might look like this `NetworkComponent networkComponent = (NetworkComponent)
         * WorkspaceFileOpener(NetworkComponent.class, new File("Net.xml"));`
         *
         * @param fileClass the type of Workpsace component to open; a subclass of WorkspaceComponent.
         * @param file      the File to open
         * @return the workspace component
         */
        fun open(fileClass: Class<*>, file: File): WorkspaceComponent {
            val extension = file.name.substring(file.name.indexOf("."))
            try {
                val method = fileClass.getMethod(
                    "open",
                    InputStream::class.java,
                    String::class.java,
                    String::class.java
                )
                val wc = method.invoke(
                    null, FileInputStream(file),
                    file.name, extension
                ) as WorkspaceComponent
                wc.currentFile = file
                wc.setChangedSinceLastSave(false)
                return wc
            } catch (e: java.lang.Exception) {
                throw java.lang.RuntimeException(e)
            }
        }
    }


}

const val BUFFER_SIZE = 1024